src/mlpack/tests/main_tests/det_test.cpp

```diff
@@ -0,0 +1,381 @@
+    // Cache in the options for this program.
+    CLI::RestoreSettings(testName);
+  }
+
+  ~DETTestFixture()
+  {
+    // Clear the settings.
+    bindings::tests::CleanMemory();
+    CLI::ClearSettings();
+  }
+};
+
+BOOST_FIXTURE_TEST_SUITE(DETMainTest, DETTestFixture);
+
+/**
+ * Check that number of output training_set_estimates and number of input data
+ * points are equal.
+ */
+BOOST_AUTO_TEST_CASE(DETOutputDimensionTest)
+{
+  arma::mat trainingData;
+  if (!data::Load("iris.csv", trainingData))
+    BOOST_FAIL("Unable to load dataset iris.csv!");
+
+  arma::mat testData;
+  if (!data::Load("iris_test.csv", testData))
+    BOOST_FAIL("Unable to load dataset iris.csv!");
+
+  // Input data.
+  SetInputParam("training", trainingData);
+  SetInputParam("test", testData);
+
+  mlpackMain();
+
+  // Check the training_set_estimates has 100 points.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("training_set_estimates").n_rows,
+                      1);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("training_set_estimates").n_cols,
+                      trainingData.n_cols);
+
+  // Check the test_set_estimates has 40 points.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("test_set_estimates").n_rows, 1);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("test_set_estimates").n_cols,
+                      testData.n_cols);
+}
+
+/**
+ * Ensure that max_leaf_size & min_leaf_size are always positive and number of
+ * folds is always non-negative.
+ */
+BOOST_AUTO_TEST_CASE(DETParamBoundTest)
+{
+  arma::mat trainingData;
+  if (!data::Load("iris.csv", trainingData))
+    BOOST_FAIL("Unable to load dataset iris.csv!");
+
+  // Test for max_leaf_size.
+
+  SetInputParam("training", trainingData);
+  SetInputParam("max_leaf_size", (int) 0);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+
+  bindings::tests::CleanMemory();
+
+  // Test for min_leaf_size.
+
+  SetInputParam("training", trainingData);
+  SetInputParam("min_leaf_size", (int) 0);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+
+  bindings::tests::CleanMemory();
+
+  // Test for folds.
+
+  SetInputParam("training", move(trainingData));
+  SetInputParam("folds", (int) -1);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Check that saved model can be reused again.
+ */
+BOOST_AUTO_TEST_CASE(DETModelReuseTest)
+{
+  arma::mat trainingData;
+  if (!data::Load("iris.csv", trainingData))
+    BOOST_FAIL("Unable to load dataset iris.csv!");
+
+  arma::mat testData;
+  if (!data::Load("iris_test.csv", testData))
+    BOOST_FAIL("Unable to load dataset iris.csv!");
+
+  // Input data.
+  SetInputParam("training", std::move(trainingData));
+  SetInputParam("test", testData);
+
+  mlpackMain();
+
+  arma::mat training_set_estimates =
```

Comment:
A minor style issue---can you avoid variable names with underscores please?  i.e. `trainingSetEstimates` not `training_set_estimates`.  https://github.com/mlpack/mlpack/wiki/DesignGuidelines#naming-conventions